src/AccountManager.mo

```diff
@@ -446,12 +471,93 @@ module {
       };
     };
 
+    func topUpRecursive(p : Principal, amount : Nat, retry : Bool) : async* TransferResponse {
+      let res = await* (
+        {
+          owner = ownPrincipal;
+          subaccount = ?Util.toSubaccount(p);
+        } |> Ledger.send(null, _, amount)
+      );
+      // catch BadFee
+      switch (res) {
+        case (#err(#BadFee { expected_fee })) {
+          let newAmount : Nat = amount - Ledger.fee() + expected_fee;
+          updateFee(expected_fee);
+          if (retry) return await* topUpRecursive(p, newAmount, false);
+        };
+        case (_) {};
+      };
+      // final return value
+      switch (res) {
+        case (#ok txid) #ok(txid);
+        case (#err err) #err(err);
+      };
+    };
+
+    func withdrawTwoStep(p : Principal, to : ICRC1.Account, amount : Nat) : async* WithdrawResponse {
+      if (amount < minimum(#withdrawal)) return #err(#TooLowQuantity);
+
+      let ?release = depositRegistry.obtainLock(p) else return #err(#NotAvailable);
+
+      var deposit = depositRegistry.get(p);
+
+      // Amount that will go to the recipient
+      let incomingAmount : Nat = amount - fee(#withdrawal);
+
+      // Sufficient amount for the user deposit account
+      let sufficientAmount : Nat = incomingAmount + Ledger.fee();
+
+      // Potential withdrawal benefit
+      var potentialBenefit : Nat = fee(#withdrawal) - Ledger.fee();
+
+      // Check whether the deposit account has the required amount for withdrawal.
+      // If not, then we top up the deposit account with the required difference.
+      if (deposit < sufficientAmount) {
+        potentialBenefit -= Ledger.fee();
+        let amountToTopUp : Nat = sufficientAmount - deposit + Ledger.fee();
+        let topUpRes = await* topUpRecursive(p, amountToTopUp, true);
```

Comment:
I don't think we can use `topUpRecursive` with `true` here. If this function goes into recursion and calls `updateFee()` then multiple values will be off. For example `sufficientAmount` was calculated based on the old ledger fee and is now wrong. `deposit` as set below in the `#ok` branch is also wrong.

In conclusion, I think we don't need a recursive version of `topUp`. If it fails we have to abort `withdrawTwoStep`.

Reply:
Maybe we need a `withdrawTwoStepRecursive` instead. If the first step (top up) fails with BadFee then we need to start over again from the beginning of the function with the `amount < minimum(#withdrawal)` test. Only the `obtainLock()` could be outside of what is repeated, it could stay locked for the second attempt.